pkg/hyper/container.go

```diff
@@ -80,23 +83,75 @@ func buildUserContainer(config *kubeapi.ContainerConfig, sandboxConfig *kubeapi.
 	}
 
 	// make volumes
-	// TODO: support adding device in upstream hyperd when creating container.
 	volumes := make([]*types.UserVolumeReference, len(config.Mounts))
 	for i, m := range config.Mounts {
 		hostPath := m.HostPath
+
 		_, volName := filepath.Split(hostPath)
-		volDetail := &types.UserVolume{
-			Name: volName + fmt.Sprintf("_%08x", rand.Uint32()),
-			// kuberuntime will set HostPath to the abs path of volume directory on host
-			Source: hostPath,
-			Format: volDriver,
+
+		// In frakti, we can both use normal container volumes (-v host:path), and also cinder-flexvolume
+		isCinderFlexvolume := false
+		// 1. host path is a directory (filter out bind mounted files)
+		if hostPathInfo, _ := os.Stat(hostPath); hostPathInfo.IsDir() {
+			// 2. tag file exists in host path
+			if _, err := os.Stat(filepath.Join(hostPath, knownflags.FlexvolumeDataFile)); !os.IsNotExist(err) {
+				// 3. then this is a cinder-flexvolume
+				isCinderFlexvolume = true
+			}
 		}
-		volumes[i] = &types.UserVolumeReference{
-			// use the generated volume name above
-			Volume:   volDetail.Name,
-			Path:     m.ContainerPath,
-			ReadOnly: m.Readonly,
-			Detail:   volDetail,
+
+		if isCinderFlexvolume {
+			// this is a cinder-flexvolume
+			optsData, err := utiljson.ReadJsonOptsFile(hostPath)
```

Comment:
Could we unmarshal the file to a structured type? so following codes don't need to do type conversions for each fields